presto-main/src/main/java/com/facebook/presto/server/TaskInfoResource.java

```diff
@@ -63,10 +95,58 @@ public TaskInfo getTaskInfo(@PathParam("taskId") TaskId taskId)
                     }
                 }
             }
-            throw new NotFoundException("TaskInfo not found for task id " + taskId.toString());
+            throw new NotFoundException("TaskInfo not found for task id " + taskId);
         }
         catch (Exception e) {
             throw new NotFoundException(e);
         }
     }
+
+    @GET
+    @Path("{taskId}")
+    @Produces(MediaType.APPLICATION_JSON)
+    public void getTaskInfo(@PathParam("taskId") TaskId taskId,
+            @QueryParam("includeLocalQueryOnly") @DefaultValue("false") boolean includeLocalQueryOnly,
+            @HeaderParam(X_FORWARDED_PROTO) String xForwardedProto,
+            @Context UriInfo uriInfo,
+            @Context HttpServletRequest servletRequest,
+            @Suspended AsyncResponse asyncResponse)
+            throws WebApplicationException
+    {
+        try {
+            if (resourceManagerEnabled && !dispatchManager.isQueryPresent(taskId.getQueryId()) && !includeLocalQueryOnly) {
+                proxyTaskInfo(servletRequest, asyncResponse, xForwardedProto, uriInfo);
+            }
+            else {
+                asyncResponse.resume(Response.ok(getTaskInfo(taskId)).build());
+            }
+        }
+        catch (NotFoundException e) {
+            asyncResponse.resume(Response.status(NOT_FOUND).build());
+        }
+    }
+
+    private void proxyTaskInfo(HttpServletRequest servletRequest, AsyncResponse asyncResponse, String xForwardedProto, UriInfo uriInfo)
+    {
+        try {
+            checkState(proxyHelper.isPresent());
+            Iterator<InternalNode> resourceManagers = internalNodeManager.getResourceManagers().iterator();
+            if (!resourceManagers.hasNext()) {
+                asyncResponse.resume(Response.status(SERVICE_UNAVAILABLE).build());
+                return;
+            }
+            InternalNode resourceManagerNode = resourceManagers.next();
+
+            URI uri = uriInfo.getRequestUriBuilder()
+                    .queryParam("includeLocalQueryOnly", true)
+                    .scheme(resourceManagerNode.getInternalUri().getScheme())
+                    .host(resourceManagerNode.getHostAndPort().toInetAddress().getHostName())
+                    .port(resourceManagerNode.getInternalUri().getPort())
+                    .build();
+            proxyHelper.get().performRequest(servletRequest, asyncResponse, uri);
+        }
```

Comment:
nit : Inverting the condition and getting rid of iterators and use Optional improves the readability.

```
            Optional<InternalNode> resourceManager = internalNodeManager.getResourceManagers().stream()
                    .findFirst();
            if (resourceManager.isPresent()) {
                InternalNode resourceManagerNode = resourceManager.get();
                URI uri = uriInfo.getRequestUriBuilder()
                        .queryParam("includeLocalQueryOnly", true)
                        .scheme(resourceManagerNode.getInternalUri().getScheme())
                        .host(resourceManagerNode.getHostAndPort().toInetAddress().getHostName())
                        .port(resourceManagerNode.getInternalUri().getPort())
                        .build();
                proxyHelper.get().performRequest(servletRequest, asyncResponse, uri);
            }
            else {
                asyncResponse.resume(Response.status(SERVICE_UNAVAILABLE).build());
            }
          
```